app/assets/javascripts/discourse/app/controllers/create-invite.js

```diff
@@ -0,0 +1,147 @@
+import Controller from "@ember/controller";
+import { action } from "@ember/object";
+import { getAbsoluteURL } from "discourse-common/lib/get-url";
+import discourseComputed from "discourse-common/utils/decorators";
+import { ajax } from "discourse/lib/ajax";
+import { extractError } from "discourse/lib/ajax-error";
+import copyText from "discourse/lib/copy-text";
+import ModalFunctionality from "discourse/mixins/modal-functionality";
+import Group from "discourse/models/group";
+import I18n from "I18n";
+
+export default Controller.extend(ModalFunctionality, {
+  onShow() {
+    let inviteKey = "";
+    for (let i = 0; i < 32; ++i) {
+      inviteKey += "0123456789abcdef"[Math.floor(Math.random() * 16)];
+    }
+
+    this.setProperties({
+      showAdvanced: false,
+      showOnly: false,
+      type: "link",
+      inviteId: null,
+      inviteKey,
+      link: getAbsoluteURL(`/invites/${inviteKey}`),
+      email: "",
+      maxRedemptionsAllowed: 1,
+      message: "",
+      topicId: null,
+      topicTitle: null,
+      groupIds: [],
+      expiresAt: moment().add(1, "week").format("YYYY-MM-DD HH:mmZ"),
+    });
+
+    Group.findAll().then((groups) => {
+      this.set("allGroups", groups.filterBy("automatic", false));
+    });
+  },
+
+  setInvite(invite) {
+    const email = invite.email || this.email;
+
+    this.setProperties({
+      error: null,
+      type: email ? "email" : "link",
+      inviteId: invite.id,
+      link: invite.link,
+      email,
+      maxRedemptionsAllowed: invite.max_redemptions_allowed,
+      message: invite.custom_message,
+      groupIds: invite.groups && invite.groups.map((g) => g.id),
+      expiresAt: invite.expires_at,
+    });
+
+    if (invite.topics && invite.topics.length > 0) {
+      this.setProperties({
+        topicId: invite.topics[0].id,
+        topicTitle: invite.topics[0].title,
+      });
+    }
+  },
+
+  @discourseComputed("expiresAt")
+  expiresAtRelative(expiresAt) {
+    return moment.duration(moment(expiresAt) - moment()).humanize();
+  },
+
+  @discourseComputed("type", "email")
+  disabled(type, email) {
+    if (type === "link") {
+    } else if (type === "email") {
+      return !email;
+    }
+  },
+
+  @discourseComputed("type", "inviteId")
+  saveLabel(type, inviteId) {
+    if (inviteId) {
+      return "user.invited.invite.update_invite";
+    } else if (type === "link") {
+      return "user.invited.invite.create_invite_link";
+    } else if (type === "email") {
+      return "user.invited.invite.send_invite_email";
+    }
+  },
+
+  @discourseComputed("type")
+  isLink(type) {
+    return type === "link";
+  },
+
+  @discourseComputed("type")
+  isEmail(type) {
+    return type === "email";
+  },
+
+  @action
+  copyLink(invite) {
+    const $copyRange = $('<p id="copy-range"></p>');
```

Comment:
Controllers shouldn't be touching the DOM. Can you move this to a component?

Reply:
I do not see how we could move this into a component. This is more of a hack to trick the browser into copying some arbitrary text (in this case, invite URL). We do this trick in other places too.

Reply:
You could create a `{{copy-link}}` component that does exactly this? It would probably only be a few lines of JS.

Reply:
Oh, I understand now. The button would be in the copy-link.

Reply:
I implemented the `{{copy-button}}` component in the latest commit.